### PR TITLE
feat: add knowledge level retrieval pipeline and MMR reranker

### PR DESCRIPTION
**Pull request by sannya-singal**
## Summary

Adds MMR (Maximal Marginal Relevance) as a retrieval strategy, along with the knowledge level pipeline it needs to work.  

`Knowledge.search()` returned the vector db's results directly, so anything that reorders results had to be implemented per adapter. Reordering also needs more candidates than the caller asked for: a document can only be surfaced if it was retrieved in the first place.   

 **Knowledge level reranking.** `Knowledge` takes an optional `reranker` that runs after the vector db returns candidates, with a widened fetch to give it something to choose between:  
                                                                                                                                                                          
  - `reranker`: applied to results before they are returned                                                                                                                         
  - `candidate_multiplier`: candidates fetched per requested result (default 5)                                                                                                        
  - `max_candidates`: ceiling on the widened fetch (default 100)                                                                                                             
                                                                                                                                                                                     
One implementation covers every adapter, including those that never implemented reranking. A reranker configured on the vector db still runs first; this runs on its output, and `Knowledge` warns when both are set.                                                                                                                               
                                                                                                                                                                                     
  **MMRReranker.** Selects results that are relevant to the query but unlike each other, so a search returns several distinct answers instead of one answer repeated. `lambda_mult` trades relevance against diversity (1.0 is pure relevance, 0.0 pure                                                                                                  
  difference).      

 Selection tracks each candidate's similarity to the nearest already selected document instead of recomputing it every iteration. 

**Vector db fixes.** MMR needs an embedding on every candidate and an embedder for the query, and neither was returned everywhere:                                                                                                                                    
                                                                                                                                                                                     
  - Cassandra, Chroma, Couchbase, Elasticsearch, OpenSearch, Pinecone and Upstash returned embeddings without the embedder that produced them. They now attach it, as the other adapters already did.                                                                                                                                                      
  - Qdrant named vector searches return a mapping of dense and sparse vectors, which was assigned to `Document.embedding` whole. Scoring raised `TypeError`, which `Knowledge` caught as a transient reranker failure, so MMR silently did nothing.                                                                                                                                          
  - Pinecone omits vectors unless asked. `PineconeDb` takes `return_vectors`, off by default so ordinary searches keep their current response size.                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                                          
  `Reranker.arerank` now runs the sync implementation in a worker thread, since a  reranker that calls a provider would otherwise block the event loop. Page backed  knowledge routes through the reranker instead of returning ahead of it.  

## Type of change

- [x] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

### Duplicate and AI-Generated PR Check

- [ ] I have searched existing [open pull requests](../../pulls) and confirmed that no other PR already addresses this issue
- [ ] If a similar PR exists, I have explained below why this PR is a better approach
- [ ] Check if this PR was entirely AI-generated (by Copilot, Claude Code, Cursor, etc.)

---

## Additional Notes

Add any important context (deployment instructions, screenshots, security considerations, etc.)

### CODE REVIEW

File: libs/agno/agno/knowledge/knowledge.py [hunk 1]
```diff
@@ -79,6 +80,17 @@ class Knowledge(RemoteKnowledge):
     page_store: Optional[Any] = None
     page_search: Optional[PageSearchConfig] = None
 
+    # Reorders results after the vector db returns them, so a strategy that needs to
+    # compare candidates against each other (diversity, recency) sees a real pool.
+    # Runs after any reranker configured on the vector db itself.
+    reranker: Optional[Reranker] = None
+    # Candidates fetched per requested result when a reranker is set. Reordering can
+    # only surface a document that was retrieved, so the pool has to exceed max_results.
+    rerank_multiplier: int = 5
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/knowledge.py]: Should the multiplier and candidates not be params of the reranker class itself?

File: libs/agno/agno/knowledge/knowledge.py [hunk 2]
```diff
@@ -948,9 +1016,9 @@ def search(
         if self.page_store is not None:
             if filters:
                 raise ValueError("Page knowledge does not support filters")
-            return self._page_documents(
-                self.search_pages(query, limit=max_results if max_results is not None else self.max_results)
-            )
+            page_limit = max_results if max_results is not None else self.max_results
+            page_documents = self._page_documents(self.search_pages(query, limit=self._search_limit(page_limit)))
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/knowledge.py]: Blocking. `PageCoordinator.search` validates `not 1 <= limit <= 20` and raises `ValueError("invalid_search_query")` before it touches the DB (`knowledge/page/_coordinator.py:1083`). The widened limit blows straight through that: with the defaults (`max_results=10`, `rerank_multiplier=5`) this asks for 50.

Repro:

```
widened limit: 50
RAISED: ValueError invalid_search_query
```

The ValueError is deliberately re-raised rather than swallowed, so it surfaces to the caller. Any `max_results >= 5` trips it, which means page knowledge plus a reranker never returns anything. Same in `asearch` at line 1073.

`test_page_store_results_are_reranked` doesn't catch it because it builds the object with `Knowledge.__new__` and stubs `search_pages` out entirely, so the real validation never runs.

Fix is a clamp to the page-search ceiling in the page branch. Worth knowing the gain is small either way: `search_pages` pops hits from the tail until the serialized result fits `max_output_bytes` (24 KB), so a widened page fetch is bounded by bytes well before it reaches 20.

File: libs/agno/agno/knowledge/reranker/mmr.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+        results: List[Document] = []
+        for index, score in selected:
+            # A shallow copy, made only so reranking_score does not land on the caller's
+            # documents: meta_data and embedding stay shared with the originals.
+            document = replace(documents[index])
+            # The MMR score at the moment this document was picked. Unlike a relevance
+            # score it is not monotonic across the list, because the candidate pool
+            # shrinks as redundancy grows: the returned order is authoritative.
+            document.reranking_score = score
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/reranker/mmr.py:126]: Blocking. `VectorSearchResult.reranking_score` is declared `ge=0.0, le=1.0` (`os/routers/knowledge/schemas.py:161`). The MMR score is `lambda*relevance - (1-lambda)*redundancy`, which goes negative whenever redundancy exceeds relevance. That is the normal case at `lambda_mult=0.5` with near-duplicate chunks, i.e. exactly what MMR is for, not an edge case.

The OS router calls `knowledge.asearch()` and then `VectorSearchResult.from_document(doc)` (`os/routers/knowledge/knowledge.py:1050,1074`), so `/knowledge/search` 500s as soon as MMR is configured:

```
[('b', 0.9992), ('a', -0.0005), ('c', -0.0024)]
b serialized ok
a FAILED: ValidationError reranking_score
c FAILED: ValidationError reranking_score
```

Relaxing the bound to `ge=-1.0` is the fix. MMR scores are genuinely in `[-1, 1]`. Side note, `SentenceTransformerReranker` writes raw cross-encoder logits into the same field, so `ge=0.0` was already too narrow. This PR just makes hitting it the default.

File: libs/agno/agno/knowledge/reranker/mmr.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        if missing:
+            # Silently returning the input order would look like MMR ran and found
+            # nothing to diversify.
+            raise ValueError(
+                "MMRReranker requires embeddings on search results, but the vector db returned "
+                f"{len(missing)} document(s) without one. Some vector dbs (Milvus, MongoDB, "
+                "Redis, Valkey) do not return embeddings on search."
+            )
+
+        limit = self.top_n if self.top_n is not None else len(documents)
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/reranker/mmr.py]: Used through `Knowledge`, `top_n` is left unset (the class docstring tells users to leave it unset), so this sets `limit = len(documents)` and MMR selects the entire pool. `_rerank_documents` then trims to `max_results` and discards the rest.

Measured on 1536-dim vectors:

| candidates | select all | select 10 |
|---|---|---|
| 50 | 0.069 s | 0.026 s |
| 100 (the ceiling) | 0.259 s | 0.053 s |

Compounding it, `MMRReranker.arerank` overrides the base and calls `_select` inline, so that 0.26 s of pure-Python CPU runs on the event loop. That undoes the reason the base `arerank` was moved to `asyncio.to_thread` in this same PR.

Passing the requested count down into `rerank()` fixes both the waste and the "leave `top_n` unset" footgun.

File: libs/agno/agno/knowledge/reranker/mmr.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+from math import sqrt
+from typing import Any, List, Optional, Sequence, Tuple
+
+from pydantic import Field, field_validator
+
+from agno.knowledge.document import Document
+from agno.knowledge.reranker.base import Reranker
+
+
+def _cosine_similarity(left: Sequence[float], right: Sequence[float]) -> float:
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/reranker/mmr.py]: Each call recomputes `left_norm` and `right_norm` from scratch, so at 100 candidates the ~5k pair comparisons redo the same ~100 norms thousands of times, roughly 3x the necessary work.

Normalising every embedding once at the top of `_select` and reducing the inner loop to a plain dot product is a small change for a 2-3x win on the hot path.

**Review by ysolanky — COMMENTED** [review]
Four findings from a pass over this, two of them blocking. Details inline.

1. Page-backed knowledge raises on every search once a reranker is set (blocking)
2. MMR scores are routinely negative, which breaks the `/knowledge/search` response schema (blocking)
3. MMR ranks the whole candidate pool and Knowledge throws most of it away, on the event loop
4. `_cosine_similarity` recomputes both norms on every pair

Separately: the README table in `cookbook/07_knowledge/02_building_blocks/` still lists only 01-06, so the four new examples need adding there.

**Review by ysolanky — COMMENTED** [review]
Re-reviewed after the latest push. All four earlier findings are addressed: the page limit is clamped to 20, `reranking_score` accepts `-1.0`, `limit` is threaded into `rerank()` and MMR's `arerank` offloads `_select` to a thread, and `_cosine_similarity` is now normalise-once plus dot product. Moving the candidate params onto `Reranker` came out well.

Three things from this round, all inline. The concurrency one on `_vector_db_reranker_suspended` is the one I would not ship as is.

File: libs/agno/agno/knowledge/knowledge.py [hunk 6]
```diff
@@ -165,6 +182,71 @@ def _page_documents(result: SearchResult) -> List[Document]:
...
+
+        Knowledge widens the fetch for its reranker, so letting the vector db reorder
+        and trim that pool first would discard the candidates it was widened for.
+        """
+        vector_db = self.vector_db
+        if self.reranker is None or vector_db is None or getattr(vector_db, "reranker", None) is None:
+            yield
+            return
+        original = vector_db.reranker
+        vector_db.reranker = None
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/knowledge.py]: This mutates an object that is shared, and holds the mutation across an `await`. Two `Knowledge` instances over one vector db is a normal setup, and so is a vector db used directly alongside a `Knowledge`. While A is inside its suspension window, anyone else reading `vector_db.reranker` sees `None`.

Repro, one shared vector db, `Knowledge` A with a knowledge-level reranker and `Knowledge` B relying on the vector db's, both searching concurrently:

```
A (has knowledge reranker): knowledge-A
B (relies on vector db)   : None      <-- expected 'vector-db'
```

B silently loses its reranking. It is timing-dependent (B has to read the attribute inside A's window), which makes it worse rather than better: it will show up under concurrent load in AgentOS and not in tests.

The guard at the top of the context manager doesn't help here, because B takes the early-return path (`self.reranker is None`) and never touches the attribute, it just reads a suspended value later.

A flag that only the owning search consults avoids the shared-state window entirely. `VectorDb` is a plain ABC and no adapter declares `reranker` at class level, so a property on the base plus a `_suppress_reranker` flag works with zero changes to the 25 adapter call sites. Passing a per-call `apply_reranker=False` down into `search()` would also work, at the cost of touching every adapter signature.

File: libs/agno/agno/knowledge/reranker/base.py [hunk 7]
```diff
@@ -10,5 +12,39 @@ class Reranker(BaseModel):
 
     model_config = ConfigDict(arbitrary_types_allowed=True, populate_by_name=True)
 
-    def rerank(self, query: str, documents: List[Document]) -> List[Document]:
+    # Candidates fetched per requested result. A reranker that scores each document on
+    # its own gains nothing from a wider pool, so the default leaves the fetch alone;
+    # one that selects a subset raises it.
+    candidate_multiplier: int = Field(default=1, ge=1)
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/reranker/base.py]: The default of 1 means no widening, and the rationale above it isn't right: a reranker that scores each document independently is precisely the case that gains from a wider pool. That is the whole retrieve-then-rerank pattern. A cross-encoder's value is that it rescues documents the bi-encoder ranked below the cutoff; scoring the same 5 the vector db already picked can only permute them.

None of `CohereReranker`, `InfinityReranker`, `SentenceTransformerReranker` or `AWSBedrockReranker` override it, so:

```
no reranker                        -> vector db asked for 5 candidates for max_results=5
scoring reranker (Cohere-shaped)   -> vector db asked for 5 candidates for max_results=5
MMRReranker                        -> vector db asked for 25 candidates for max_results=5
```

So telling users to move their Cohere reranker from the vector db to `Knowledge` gets them the same order-only behaviour they had before, unless they also know to set `candidate_multiplier`. That undercuts the migration this PR is setting up.

I'd rather see the base default at 3-5 and let a reranker that genuinely doesn't benefit opt down to 1, than have the four shipped rerankers silently no-op. If the concern is the cost increase landing on existing users, that is worth calling out in the changelog rather than defaulting the feature off.

File: libs/agno/agno/knowledge/knowledge.py [hunk 8]
```diff
@@ -165,6 +182,71 @@ def _page_documents(result: SearchResult) -> List[Document]:
...
+        """Widen the vector db fetch so the reranker has candidates to choose between."""
+        if self.reranker is None:
+            return max_results
+        # The reranker decides how wide its own pool needs to be.
+        return self.reranker.search_limit(max_results)
+
+    def _rerank_documents(self, query: str, documents: List[Document], max_results: int) -> List[Document]:
+        """Apply the knowledge-level reranker, then trim to the caller's requested count."""
+        if self.reranker is None:
+            return documents[:max_results]
```
Comment (ysolanky) [inline libs/agno/agno/knowledge/knowledge.py]: Minor, and the only one of these that touches the default path. `_rerank_documents` now runs on every search, and the no-reranker branch slices rather than passing through, so every existing `Knowledge.search()` result is truncated where previously it was returned as the adapter produced it.

I checked the adapters and they all pass `limit` down to the backend as `top_k` / `size` / `LIMIT`, so this should be a no-op in practice. But it is an unconditional change in the path every current user is on, which is where you least want one. `if self.reranker is None: return documents` keeps the guarantee for free.

File: cookbook/07_knowledge/02_building_blocks/08_mmr_diverse_results.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        candidate_multiplier=5,
+        # Ceiling on that widened fetch, whatever max_results is asked for.
+        max_candidates=100,
+    ),
+)
+
+agent = Agent(
+    model=OpenAIResponses(id="gpt-5.6-luna"),
+    knowledge=knowledge,
+    search_knowledge=True,
```
Comment (kausmeows) [inline cookbook/07_knowledge/02_building_blocks/08_mmr_diverse_results.py]: this is true by default in agno, so can skip it here

File: libs/agno/agno/knowledge/reranker/mmr.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
```
Comment (kausmeows) [inline libs/agno/agno/knowledge/reranker/mmr.py]: maybe rather have this as a general public utility function in a module somewhere? as i'm assuming it will be frequently used across the sdk and in future rerankers too, so would be a cleaner import that way.

File: libs/agno/agno/knowledge/reranker/mmr.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+def _unit(vector: Sequence[float]) -> List[float]:
+    """Scale to unit length so similarity reduces to a dot product."""
+    norm = sqrt(sum(value * value for value in vector))
+    if norm <= 0.0:
+        return [0.0] * len(vector)
+    return [value / norm for value in vector]
+
+
+def _dot(left: Sequence[float], right: Sequence[float]) -> float:
```
Comment (kausmeows) [inline libs/agno/agno/knowledge/reranker/mmr.py]: same for this.

File: libs/agno/agno/knowledge/knowledge.py [hunk 12]
```diff
@@ -117,6 +127,13 @@ def __init__(
         self.embedding_retry_backoff = embedding_retry_backoff
         self.page_store = page_store
         self.page_search = page_search
+        self.reranker = reranker
+        if reranker is not None and getattr(vector_db, "reranker", None) is not None:
+            log_warning(
```
Comment (kausmeows) [inline libs/agno/agno/knowledge/knowledge.py:132]: should we also mention that the vector db one will be deprecated in future so use this?

Comment (sannya-singal) [reply libs/agno/agno/knowledge/knowledge.py:132]: this is excluded intentionally in this PR, will create a follow-up PR to add a deprecation warning along with bumping the cookbooks, tests, etc. to knowledge level.

Comment (sannya-singal) [reply libs/agno/agno/knowledge/knowledge.py:132]: I will soften the warning to "prefer the one on Knowledge" rather than making a deprecation note.

**Review by ysolanky — APPROVED** [review]
